File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int len(ListNode* head){
        int cnt=0;
        ListNode* temp=head;
        while(temp){
            cnt++;
            temp=temp->next;
        }
        return cnt;
    }
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int l=len(head);
        l-=n;
        if(l==0) return head->next;
        int cnt=0;
        ListNode* temp=head;
        while(temp){
            cnt++;
            if(cnt==l) temp->next=temp->next->next;
            temp=temp->next;
        }
        return head;
    }
};

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int countNodes(ListNode head){
        int cnt=0;
        ListNode ptr=head;
        while(ptr!=null){
            ptr=ptr.next;
            cnt++;
        }
        return cnt;
    }
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int cnt=countNodes(head);
        int rem=cnt-n;
        if (rem == 0) {
            return head.next;
        }
        ListNode ptr=head;
        for (int i = 1; i < rem; i++) {
            ptr = ptr.next;
        }
        if (ptr.next != null) {
            ptr.next = ptr.next.next;
        }
        return head;
    }
}